src/common/inc/fnwatchdog.h

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+
+#pragma once
+
+#include "pooltag.h"
+
+DECLARE_HANDLE(FN_WATCHDOG_HANDLE);
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+VOID
+FN_WATCHDOG_CALLBACK(
+    _In_ VOID *CallbackContext
+    );
+
+typedef struct _FN_WATCHDOG {
+    FN_WATCHDOG_CALLBACK *Callback;
+    VOID *CallbackContext;
+    HANDLE ThreadHandle;
+    KEVENT CloseEvent;
+    LARGE_INTEGER Timeout;
+    UINT32 IntervalMs;
+} FN_WATCHDOG;
+
+inline
+_IRQL_requires_same_
+_Function_class_(KSTART_ROUTINE)
+VOID
+FnWatchdogWorker(
+    _In_ VOID *Context
+    )
+{
+    FN_WATCHDOG *Watchdog = Context;
+    NTSTATUS Status;
+
+    do {
+        Watchdog->Timeout.QuadPart += RTL_MILLISEC_TO_100NANOSEC(Watchdog->IntervalMs);
+
+        Status =
+            KeWaitForSingleObject(
+                &Watchdog->CloseEvent, Executive, KernelMode, FALSE, &Watchdog->Timeout);
+
+        Watchdog->Callback(Watchdog->CallbackContext);
```

Comment:
I would have expected the watchdog callback to only occur it the timeout was reached?

Reply:
Yeah, really this file is just a periodic timer. I should rename it.